Memory manager support for Windows nodes

marosset opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

/kind feature

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation


Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

This PR adds support for memory manager on Windows nodes and add a new `BestEffort` policy which allows giving preferences to the Windows OS for which CPUs to bind containers to.
Windows cannot guarantee scheduling workloads to specific NUMA nodes so instead we query the OS for which CPUs are part of a given NUMA node and specify that when starting containers.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Part of https://github.com/kubernetes/kubernetes/issues/125262

To enable memory manager support the kubelet configuration would look something like

```toml
cpuManagerPolicy: static
memoryManagerPolicy: BestEffort
systemReserved: 
  cpu: 500m
  memory: 500Mi
featureGates:
  MemoryManager: true
  WindowsCPUAndMemoryAffinity: true

```
#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Added Windows support for the node memory manager.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-windows/4885-windows-cpu-and-memory-affinity
```

/sig windows node
/milestone v1.32
/area kubelet
/assign @jsturtevant

marosset committed:
Memory manager support for Windows nodes

ffromani commented on pkg/kubelet/cm/container_manager_windows.go:
```diff
@@ -93,12 +95,21 @@ func (cm *containerManagerImpl) Start(ctx context.Context, node *v1.Node,
 		}
+
+		// Initialize memory manager
+		if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.MemoryManager) {
```
please rebase as we merged https://github.com/kubernetes/kubernetes/pull/128517

ffromani commented on pkg/kubelet/cm/container_manager_windows.go:
```diff
@@ -93,12 +95,21 @@ func (cm *containerManagerImpl) Start(ctx context.Context, node *v1.Node,
+
+		// Initialize memory manager
+		if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.MemoryManager) {
+			containerMap, _ := buildContainerMapAndRunningSetFromRuntime(ctx, runtimeService)
```
you can now use the value returned on line 96 above, like the linux container manager does.

ffromani commented on pkg/kubelet/cm/internal_container_lifecycle_windows.go:
```diff
@@ -20,30 +20,160 @@ limitations under the License.
 }
+
+// computeCPUSet returns a map of CPU IDs to an empty struct based on the provided group affinities
+func computeCPUSet(affinities []winstats.GroupAffinity) map[int]struct{} {
```
can we use sets from `apimachinery/util` (or its most current replacement like perhaps `k8s.io/utils`)?

marosset replied:
yes, i'll make some updates

marosset replied:
Switched to apimachinery/util since it had everything I needed.

ffromani commented on pkg/kubelet/cm/memorymanager/memory_manager.go:
```diff
@@ -150,6 +155,13 @@ func NewManager(policyName string, machineInfo *cadvisorapi.MachineInfo, nodeAll
 		}
 
+	case policyTypeBestEffort:
+		if runtime.GOOS == "windows" {
+
+		} else {
+			return nil, fmt.Errorf("policy \"%s\" is not avaiable for platform \"%s\"", policyTypeBestEffort, runtime.GOOS)
+		}
```
```
if runtime.GOOS != "windows" {
    return nil, fmt.Errorf("policy \"%s\" is not avaiable for platform \"%s\"", policyTypeBestEffort, runtime.GOOS)
}
```
?

marosset replied:
oops

ffromani commented on pkg/kubelet/cm/memorymanager/policy_best_effort.go:
```diff
@@ -0,0 +1,75 @@
+
+const policyTypeBestEffort policyType = "BestEffort"
+
+// bestEffortPolicy is implementation of the policy interfact for the BestEffort policy
```
interface?

ffromani commented on pkg/kubelet/cm/memorymanager/policy_best_effort.go:
```diff
@@ -0,0 +1,75 @@
+
+// bestEffortPolicy is implementation of the policy interfact for the BestEffort policy
+type bestEffortPolicy struct {
+	static *staticPolicy
```
Is not obvious to me what do we gain by wrapping the static policy. Could you please elaborate a bit?

marosset replied:
For Windows we want the new `BestEffort` policy to have the same logic as the existing `static` policy.
I figured that doing this would reduce a lot of code duplication.

I can update the comments here to reflect that or maybe I can have the kubelet arguments specify `BestEffortPolicy` and then create a `staticPolicy` instance here for Windows.

Do you all have a preference @ffromani @jsturtevant ?

jsturtevant replied:
Updating comments and wrapping seems to make sense to me. This will allow to adjust more easily if we learn of reasons to adjust.

ffromani replied:
I'll need to refresh my memory over what we discussed in the KEP before I can answer. Right now I can surely say comments is a great starting point.

ffromani replied:
ok, first: I've nothing against reusing the code like this. We didn't do often in the codebase, but why not?
The issue I see however is that from my interpretation of the KEP I would have expected different hint generation on windows based on https://github.com/kubernetes/enhancements/tree/master/keps/sig-windows/4885-windows-cpu-and-memory-affinity#windows-memory-considerations

I think we can settle this with a good set of testcases, at very least unit tests. which seems lacking to me?
Granted, the policy is actually a thin wrapper over the static policy, which has its own tests; the tests I recommend to add are to ensure and document the expected behavior on windows.

The tests should cover preferably also the topology manager merge hint process. It will be similar in spirit to what you are adding to `internal_container_lifecycle_windows_test.go`, but covering the generic topology manager flow.

There's likely not enough time to do this for this cycle. Plus I want to be practical: it's an alpha feature, disabled by default, and the early feedback from having some functionality in is important as well. So I don't think it is blocking, but I _DO_ think this is **beta**-blocking (to discuss this topic in detail at least) and possibly we can need another alpha iteration to sort this out depending on the aforementioned feedback

jsturtevant replied:
> The issue I see however is that from my interpretation of the KEP I would have expected different hint generation on windows based on https://github.com/kubernetes/enhancements/tree/master/keps/sig-windows/4885-windows-cpu-and-memory-affinity#windows-memory-considerations

I didn't see a difference in what we would provide for a hint. The major difference is in the semantic meaning of the hint, which on Windows we can't guarantee that it the hint will be respected, only make that suggestion.  What kind of difference are you thinking?

>There's likely not enough time to do this for this cycle. Plus I want to be practical: it's an alpha feature, disabled by default, and the early feedback from having some functionality in is important as well. So I don't think it is blocking, but I DO think this is beta-blocking (to discuss this topic in detail at least) and possibly we can need another alpha iteration to sort this out depending on the aforementioned feedback

I appreciate this sentiment.  We do have a user we are working with that is willing to provide feedback but is looking for it to be in an Alpha release.  From my understanding of the challenges on Linux, these types of features need some real feedback to get them dialed in, and I anticipate we will get feedback that we will need to incorporate back into the design.

ffromani replied:
> > The issue I see however is that from my interpretation of the KEP I would have expected different hint generation on windows based on https://github.com/kubernetes/enhancements/tree/master/keps/sig-windows/4885-windows-cpu-and-memory-affinity#windows-memory-considerations
> 
> I didn't see a difference in what we would provide for a hint. The major difference is in the semantic meaning of the hint, which on Windows we can't guarantee that it the hint will be respected, only make that suggestion. What kind of difference are you thinking?

I'm basing my expectations on what I saw so far on linux for the cpumanager policy. I would like to see if for example returning a different set of NUMA bitmasks or different values for `Preferred` field.

If we add a thin wrapper over the Static policy, we are saying that this policy should behave 1:1 like the static policy, and we are unnecessarily coupling the behavior. What I would rather see is a specification (and unit tests are a good vehicle for that) of the expected behavior, more precise than (yes, I'm oversimplifying here) "whatever Static policy does, but not enforcing".

Lacking specification is something that biten us already multiple already, so we should IMO go in the direction to have the spec spelled out more precisely, and the testcase well defined. If, after that, we can reuse the implementation, great.

If nothing else (which I can accept!) this should be discussed and elaborated in the KEP design details.
I'm happy to elaborate further in the next cycle. Again, more details in the KEP to elaborate more precisely the behavior can possibly be all we need.

> > There's likely not enough time to do this for this cycle. Plus I want to be practical: it's an alpha feature, disabled by default, and the early feedback from having some functionality in is important as well. So I don't think it is blocking, but I DO think this is beta-blocking (to discuss this topic in detail at least) and possibly we can need another alpha iteration to sort this out depending on the aforementioned feedback
> 
> I appreciate this sentiment. We do have a user we are working with that is willing to provide feedback but is looking for it to be in an Alpha release. From my understanding of the challenges on Linux, these types of features need some real feedback to get them dialed in, and I anticipate we will get feedback that we will need to incorporate back into the design.

Fine, so we will evaluate during the 1.33 cycle if we can move to beta (taking into account my oen feedback above) or if we need another iteration.

The rest of the code overall LGTM, I'll have another pass ASAP but from my recollections I don't have anything major on my notes.

ffromani commented:
/triage accepted
/priority important-soon

initial priority assessment subject to be changed

jsturtevant commented on pkg/kubelet/cm/internal_container_lifecycle_windows.go:
```diff
@@ -20,30 +20,160 @@ limitations under the License.
+}
+
+// convertToGroupAffinities converts a cpuset.CPUSet to a slice of winstats.GROUP_AFFINITY
+func convertToGroupAffinities(cpuSet cpuset.CPUSet) []winstats.GroupAffinity {
```
This looks pretty close to https://github.com/kubernetes/kubernetes/blob/3140807126c86163584974e8a0835c2f5907af97/pkg/kubelet/winstats/cpu_topology.go#L91  

Could we provide another helper function and/or adjust the logic to convert to affinities?  My concern would be spreading the logic for calculating back and forth between kubelet's representation  (integers for the CPU's) to Windows representation of groupAffinities.

marosset replied:
I refactored some of the logic and got rid of the need for this method

jsturtevant commented on pkg/kubelet/cm/internal_container_lifecycle_windows.go:
```diff
@@ -20,30 +20,160 @@ limitations under the License.
+	// set affinity based on available managers
+	var finalCPUSet map[int]struct{}
+
+	if i.cpuManager != nil && !allocatedCPUs.IsEmpty() && i.memoryManager != nil && numaNodes.Len() > 0 {
```
Thanks for all the unit tests on the merging of sets.  There is quite a bit of logic here, that might be nice to have unit test to verify all this behavior (the three scenarios and with each manager on/off)

marosset replied:
agree, i'll try and add some

marosset replied:
I did some slight refactoring and added unit tests for this logic - PTAL

jsturtevant replied:
Thanks!

ffromani commented:
@marosset adding the policy name ultimates surfaces to the user and to the kubeletconfig API object so this change *MAY* review api-review, please take this into account.

sftim commented:
Changelog suggestion
```diff
-Windows: Support memory manager on Windows 
+Added Windows support for the node memory manager.
```

ffromani commented on pkg/kubelet/cm/internal_container_lifecycle_windows.go:
```diff
@@ -20,30 +20,119 @@ limitations under the License.
+		return nil
+	}
+
+	klog.Info("PreCreateContainer for Windows")
```
you may want to set this to `V(2)` at least, likely `V(4)` otherwise it can get spammy very fast

marosset replied:
updated to v(4)

ffromani commented on pkg/kubelet/cm/memorymanager/memory_manager.go:
```diff
@@ -150,6 +155,20 @@ func NewManager(policyName string, machineInfo *cadvisorapi.MachineInfo, nodeAll
 		}
 
+	case policyTypeBestEffort:
+		if runtime.GOOS == "windows" {
+			systemReserved, err := getSystemReservedMemory(machineInfo, nodeAllocatableReservation, reservedMemory)
```
nit: you can do like `Static` above:
```
if runtime.GOOS != "windows" {
    return nil, fmt.Errorf("policy \"%s\" is not available for platform \"%s\"", policyTypeBestEffort, runtime.GOOS)
}
systemReserved, err := ...
```

ffromani commented on pkg/kubelet/cm/memorymanager/memory_manager.go:
```diff
@@ -140,6 +141,10 @@ func NewManager(policyName string, machineInfo *cadvisorapi.MachineInfo, nodeAll
 
 	case policyTypeStatic:
+		if runtime.GOOS == "windows" {
+			return nil, fmt.Errorf("policy \"%s\" is not not available on Windows", policyTypeStatic)
```
nit: can we use `%q` instead of `\"%s\"` ?

jsturtevant commented on pkg/kubelet/cm/memorymanager/policy_best_effort.go:
```diff
@@ -0,0 +1,80 @@
+)
+
+// On Windows we want to use the same logic as the StaticPolicy to compute the memory topology hints
+// but unlike linux based systems, on Windows systesm numa nodes cannot be directly assigned or guaranteed via Windows APIs
```
```suggestion
// but unlike linux based systems, on Windows systems numa nodes cannot be directly assigned or guaranteed via Windows APIs
```

jsturtevant commented on pkg/kubelet/cm/memorymanager/policy_best_effort.go:
```diff
@@ -0,0 +1,80 @@
+
+// On Windows we want to use the same logic as the StaticPolicy to compute the memory topology hints
+// but unlike linux based systems, on Windows systesm numa nodes cannot be directly assigned or guaranteed via Windows APIs
+// (but they are usually respoected). Because of this we don't want to have users specify "StaticPolicy" for the memory manager
```
```suggestion
// (windows scheduler will use the numa node that is closest to the cpu assigned therefor respecting the numa node assignment as a best effort). Because of this we don't want to have users specify "StaticPolicy" for the memory manager
```

or something like that?

ffromani reviewed (commented):
LGTM once comments have been addressed.

I think we should *try* to reconcile in the next cycle the windows-specific alignment code with the generic code from topology manager. IOW, instead of asking cpumanager for the affinities, we should try to query topology manager, which is supposed to be the source of truth for alignment.

TL;DR: the windows-specific code is reimplementing a narrowed-down version of topology manager, and we should try to converge back in topology manager in the future.

[EDIT] I'm aware the linux container internal container lifecycle code does the same; but it's also true that up until now that was the only true implementation. Now that we have two, we should give an honest attempt to generalize. Worst case we keep the _status quo_ with comments explaining why.

marosset commented:
> Changelog suggestion
> 
> ```diff
> -Windows: Support memory manager on Windows 
> +Added Windows support for the node memory manager.
> ```

Done

marosset commented:
thanks @ffromani 
I think I addressed all your feedback - please take another look when you get a chance

I'll open up a separate PR to include unit testing for the BestEffort policy as a `beta` graduation requirement to the KEP.

marosset commented:
> @marosset adding the policy name ultimates surfaces to the user and to the kubeletconfig API object so this change _MAY_ require api-review, please take this into account.

@ffromani - Do you know who would be able to make that decision?

marosset commented:
/assign @mrunalp @dchen1107 for node approval

marosset commented:
> > @marosset adding the policy name ultimates surfaces to the user and to the kubeletconfig API object so this change _MAY_ require api-review, please take this into account.
> 
> @ffromani - Do you know who would be able to make that decision?

/label api-review
In case this does...

marosset commented:
It looks like there are some unrelated infra issue with the windows / CAPZ test passes - please ignore those failure for now.

jsturtevant commented:
/restest

jsturtevant commented:
/test pull-kubernetes-e2e-capz-windows-master

jsturtevant commented:
/test pull-kubernetes-e2e-capz-windows-master

jsturtevant commented:
We are still recovering from an issue with the azure sub
/test pull-kubernetes-e2e-capz-windows-master

jsturtevant commented:
I've pulled this branch and run a few different scenarios, and I am seeing the expected behavior.  A few of the new unit tests did fail but they looked like typos nothing serious.  Once those are resolved this lgtm.

marosset commented:
> I've pulled this branch and run a few different scenarios, and I am seeing the expected behavior. A few of the new unit tests did fail but they looked like typos nothing serious. Once those are resolved this lgtm.

The unit test caught an issue that got introduced by a small refactor I made a few commits ago
https://github.com/kubernetes/kubernetes/pull/128560/files#diff-6711ce20b4cecd0fa883ddc3f1c6dde5c56dcfebca347b4de3e6e794feab5269R111 was incorrect.

All the unit tests are passing on Windows now.
Thanks for pointing this out @jsturtevant

marosset commented:
/label tide/merge-method-squash

ffromani reviewed (commented):
/lgtm

ffromani commented:
@marosset / @jsturtevant is not clear if this PR is _needed_ for https://github.com/kubernetes/enhancements/issues/4885 (I'm fine in both cases per my comments above). If so, can you please update the list on the issue? Thanks!

ffromani commented:
folks kindly be aware of https://github.com/kubernetes/kubernetes/pull/128657

ffromani commented:
/lgtm cancel

need to rebase on top of https://github.com/kubernetes/kubernetes/pull/128657

marosset commented:
> /lgtm cancel
> 
> need to rebase on top of #128657

Ack - I'll rebase on that as soon as it merges.

ffromani reviewed (commented):
/lgtm

thanks for the rebase and sorry for the churn. You use the recently added `containerMap.Clone()`, so looks fine again to me.

mrunalp commented on pkg/kubelet/cm/memorymanager/memory_manager.go:
```diff
@@ -140,6 +141,10 @@ func NewManager(policyName string, machineInfo *cadvisorapi.MachineInfo, nodeAll
 
 	case policyTypeStatic:
+		if runtime.GOOS == "windows" {
+			return nil, fmt.Errorf("policy %q is not not available on Windows", policyTypeStatic)
```
nit: typo: `not not`

marosset replied:
fixed

ffromani commented:
beta-blocking (or perhaps as later bugfix?) we need to update `cmd/kubelet/app/options/options.go`

mrunalp commented:
/lgtm

jsturtevant commented:
/lgtm

tengqm commented:
Do we have documentation updates (including the feature gate) to the website?

marosset commented:
> Do we have documentation updates (including the feature gate) to the website?

I we'll use https://github.com/kubernetes/website/pull/48469 to update docs for both cpu manager (https://github.com/kubernetes/kubernetes/pull/125296) and memory manager (this PR) support for windows.
